### PR TITLE
Cached stable hash cleanups

### PR DESCRIPTION
**Pull request by oli-obk**
r? @nnethercote 

Add a sanity assertion in debug mode to check that the cached hashes are actually the ones we get if we compute the hash each time.

Add a new data structure that bundles all the hash-caching work to make it easier to re-use it for different interned data structures

### CODE REVIEW

File: compiler/rustc_data_structures/src/intern.rs [hunk 1]
```diff
@@ -108,5 +110,76 @@ where
     }
 }
 
+/// A helper trait so that `Interned` things can cache stable hashes reproducibly.
+pub trait InternedHashingContext {
+    fn with_def_path_and_no_spans(&mut self, f: impl FnOnce(&mut Self));
+}
+
+#[derive(Copy, Clone)]
+pub struct InTy<T> {
```
Comment (nnethercote) [inline compiler/rustc_data_structures/src/intern.rs:119]: What's the meaning of the name `InTy`? It's not obviously meaningful to me.

Comment (nnethercote) [reply compiler/rustc_data_structures/src/intern.rs:119]: Feels like there should be a comment on this type, too.

Comment (nnethercote) [reply compiler/rustc_data_structures/src/intern.rs:119]: Also, is this `InTy` wrapping likely to occur with other types? The generic `T` here suggests yes.

Comment (oli-obk) [reply compiler/rustc_data_structures/src/intern.rs]: I documented and renamed this.

**Comment by nnethercote** [comment]
The new name helps, thanks.

I'm still a little unclear about the goal here.
- I can see it's nice to pull the hash out of `TyS`. Because the hash is auxiliary, rather than being a fundamental part of the type.
- The PR description says "I pulled out the perf improvement from https://github.com/rust-lang/rust/pull/94487" but the perf results are neutral. Is that expected?

File: compiler/rustc_middle/src/ty/mod.rs [hunk 2]
```diff
@@ -433,32 +433,31 @@ crate struct TyS<'tcx> {
...
-
-    /// The stable hash of the type. This way hashing of types will not have to work
-    /// on the address of the type anymore, but can instead just read this field
-    stable_hash: Fingerprint,
 }
 
 // `TyS` is used a lot. Make sure it doesn't unintentionally get bigger.
 #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
-static_assert_size!(TyS<'_>, 56);
+static_assert_size!(TyS<'_>, 40);
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/mod.rs:441]: Can you check TypeAndStableHash<TyS> too?

File: compiler/rustc_data_structures/src/intern.rs [hunk 3]
```diff
@@ -108,5 +110,87 @@ where
...
+
+/// A helper type that you can wrap round your own type in order to automatically
+/// cache the stable hash on creation and not recompute it whenever the stable hash
+/// of the type is computed.
+/// This is only done in incremental mode. You can also opt out of caching by using
+/// StableHash::ZERO for the hash, in which case the hash gets computed each time.
+/// This is useful if you have values that you intern but never (can?) use for stable
+/// hashing.
+#[derive(Copy, Clone)]
+pub struct TypeAndStableHash<T> {
```
Comment (bjorn3) [inline compiler/rustc_data_structures/src/intern.rs:126]: Maybe call it WithStableHash?

**Comment by bjorn3** [comment]
I think @oli-obk means that the perf improvement is not included in this PR and only the refactorings that enable the perf improvement change.

**Comment by oli-obk** [comment]
Nope, the other PR had a perf improvement for these commits, but that seems to be gone now, maybe it was spurious or randomly dependent on inlining decisions

**Comment by nnethercote** [comment]
So is this still worthwhile for the conceptual improvement of separating the hash from `TyS`?

**Comment by oli-obk** [comment]
Yes, I can investigate perf things independently, editing the main message now to reflect that (and will address bjorn's comments)

**Comment by oli-obk** [comment]
Renamed and force pushed

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 4]
```diff
@@ -1266,18 +1266,29 @@ pub trait PrettyPrinter<'tcx>:
             ty::Ref(
                 _,
                 Ty(Interned(
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1268]: Maybe use nested match with the inner match matching on `pointee.kind()`? This indentation is getting crazy and pretty unreadable.

**Comment by nnethercote** [comment]
Ok by me, I'll let you address @bjorn3's final comment if you want.

@bors delegate+

**Pull request merged**